src/pudl/metadata/classes.py

```diff
@@ -0,0 +1,911 @@
+    To use them in a class, use an underscore prefix and an alias.
+
+    Examples:
+        >>> class Class(Base):
+        ...     fields_: List[str] = pydantic.Field(alias="fields")
+        >>> m = Class(fields=['x'])
+        >>> m
+        Class(fields=['x'])
+        >>> m.fields
+        ['x']
+        >>> m.fields = ['y']
+        >>> m.dict()
+        {'fields': ['y']}
+    """
+
+    class Config:
+        """Custom Pydantic configuration."""
+
+        validate_all: bool = True
+        validate_assignment: bool = True
+        extra: str = 'forbid'
+
+    def dict(self, *args, by_alias=True, **kwargs) -> dict:  # noqa: A003
+        """Return as a dictionary."""
+        return super().dict(*args, by_alias=by_alias, **kwargs)
+
+    def json(self, *args, by_alias=True, **kwargs) -> str:
+        """Return as JSON."""
+        return super().json(*args, by_alias=by_alias, **kwargs)
+
+    def __getattribute__(self, name: str) -> Any:
+        """Get attribute."""
+        if name in ("fields", "schema") and f"{name}_" in self.__dict__:
+            name = f"{name}_"
+        return super().__getattribute__(name)
+
+    def __setattr__(self, name, value) -> None:
+        """Set attribute."""
+        if name in ("fields", "schema") and f"{name}_" in self.__dict__:
+            name = f"{name}_"
+        super().__setattr__(name, value)
+
+    def __repr_args__(self) -> List[Tuple[str, Any]]:
+        """Returns the attributes to show in __str__, __repr__, and __pretty__."""
+        return [
+            (a[:-1] if a in ("fields_", "schema_") else a, v)
+            for a, v in self.__dict__.items()
+        ]
+
+
+# ---- Class attribute types ---- #
+
+
+String = pydantic.constr(min_length=1, strict=True)
+"""
+Non-empty :class:`str` (anything except "").
+"""
+
+Bool = pydantic.StrictBool
+"""
+Any :class:`bool` (`True` or `False`).
+"""
+
+Float = pydantic.confloat(ge=0, strict=True)
+"""
+Positive :class:`float`.
+"""
+
+
+def StrictList(item_type: Type = Any) -> pydantic.ConstrainedList:  # noqa: N802
+    """
+    Non-empty :class:`list`.
+
+    Allows :class:`list`, :class:`tuple`, :class:`set`, :class:`frozenset`,
+    :class:`collections.deque`, or generators and casts to a :class:`list`.
+    """
+    return pydantic.conlist(item_type=item_type, min_items=1)
+
+
+# ---- Class attribute validators ---- #
+
+
+def _check_unique(value: list = None) -> Optional[list]:
+    """Check that input list has unique values."""
+    if value:
+        for i in range(len(value)):
+            if value[i] in value[:i]:
+                raise ValueError(f"contains duplicate {value[i]}")
+    return value
+
+
+def _stringify(value: Any = None) -> Optional[str]:
+    """Convert input to string."""
+    if value:
+        return str(value)
+    return value
+
+
+def _validator(*names, fn: Callable) -> Callable:
+    """
+    Construct reusable Pydantic validator.
```

Comment:
This is a helper function that allows us to use a custom function as a validator in multiple models with minimal markup.
See https://pydantic-docs.helpmanual.io/usage/validators/#reuse-validators.

Note that validators are not limited to validation (i.e. raise an error) but can also modify the value. So in that spirit, `_stringify` is being used to convert value to string. This is used where I am using e.g. `email: pydantic.EmailStr` to validate `email` as a valid email. In the process, the email is cast to type `EmailStr` and stored in the model as such. Thinking this would be confusing, I add `_stringify` to ensure it is instead stored in the model as a plain `str`.